ee/clickhouse/sql/cohort.py

```diff
@@ -22,45 +22,49 @@
 )
 
 TRUNCATE_COHORTPEOPLE_TABLE_SQL = f"TRUNCATE TABLE IF EXISTS cohortpeople ON CLUSTER '{CLICKHOUSE_CLUSTER}'"
-DROP_COHORTPEOPLE_TABLE_SQL = f"DROP TABLE IF EXISTS cohortpeople ON CLUSTER '{CLICKHOUSE_CLUSTER}'"
 
-REMOVE_PEOPLE_NOT_MATCHING_COHORT_ID_SQL = """
-INSERT INTO cohortpeople
-SELECT person_id, cohort_id, %(team_id)s as team_id,  -1 as _sign
-FROM cohortpeople
-JOIN (
-    SELECT id, argMax(properties, person._timestamp) as properties, sum(is_deleted) as is_deleted FROM person WHERE team_id = %(team_id)s GROUP BY id
-) as person ON (person.id = cohortpeople.person_id)
-WHERE cohort_id = %(cohort_id)s
-AND
-    (
-        person.is_deleted = 1 OR NOT person_id IN ({cohort_filter})
-    )
-"""
+
+COHORTPEOPLE2_TABLE_ENGINE = lambda: VersionedCollapsingMergeTree("cohortpeople2", sign="sign", ver="version")
+CREATE_COHORTPEOPLE2_TABLE_SQL = lambda: """
+CREATE TABLE IF NOT EXISTS cohortpeople2 ON CLUSTER '{cluster}'
+(
+    person_id UUID,
+    cohort_id Int64,
+    team_id Int64,
+    sign Int8,
+    version UInt64
+) ENGINE = {engine}
+Order By (team_id, cohort_id, person_id)
+{storage_policy}
+""".format(
+    cluster=CLICKHOUSE_CLUSTER, engine=COHORTPEOPLE2_TABLE_ENGINE(), storage_policy="",
+)
+
+TRUNCATE_COHORTPEOPLE2_TABLE_SQL = f"TRUNCATE TABLE IF EXISTS cohortpeople2 ON CLUSTER '{CLICKHOUSE_CLUSTER}'"
 
 GET_COHORT_SIZE_SQL = """
 SELECT count(*)
 FROM (
     SELECT 1
-    FROM cohortpeople
-    WHERE team_id = %(team_id)s AND cohort_id = %(cohort_id)s
+    FROM cohortpeople2
+    WHERE team_id = %(team_id)s AND cohort_id = %(cohort_id)s AND version = %(version)s
     GROUP BY person_id, cohort_id, team_id
     HAVING sum(sign) > 0
 )
 """
 
-INSERT_PEOPLE_MATCHING_COHORT_ID_SQL = """
-INSERT INTO cohortpeople
-    SELECT id, %(cohort_id)s as cohort_id, %(team_id)s as team_id, 1 as _sign
-    FROM (
-        SELECT id, argMax(properties, person._timestamp) as properties, sum(is_deleted) as is_deleted FROM person WHERE team_id = %(team_id)s GROUP BY id
-    ) as person
-    LEFT JOIN (
-        SELECT person_id, sum(sign) AS sign FROM cohortpeople WHERE cohort_id = %(cohort_id)s AND team_id = %(team_id)s GROUP BY person_id
-    ) as cohortpeople ON (person.id = cohortpeople.person_id)
-    WHERE (cohortpeople.person_id = '00000000-0000-0000-0000-000000000000' OR sign = 0)
-    AND person.is_deleted = 0
-    AND id IN ({cohort_filter})
+RECALCULATE_COHORT_BY_ID = """
+INSERT INTO cohortpeople2
+SELECT id, %(cohort_id)s as cohort_id, %(team_id)s as team_id, 1 AS sign, %(new_version)s AS version
+FROM (
+    SELECT id, argMax(properties, person._timestamp) as properties, sum(is_deleted) as is_deleted FROM person WHERE team_id = %(team_id)s GROUP BY id
+) as person
+WHERE person.is_deleted = 0
+AND id IN ({cohort_filter})
+UNION ALL
+SELECT person_id, cohort_id, team_id, -1 as sign, version
```

Comment:
Do we need to do the `sum(sign)` aggregation here too? Seems like it would be a best practive to avoid over-deleting a row (although, I'm not sure that matters anymore?)